crates/walrus-service/src/node/events/event_blob_writer.rs

```diff
@@ -563,17 +566,38 @@ impl EventBlobWriterFactory {
             consecutive_uncertified,
             num_uncertified_blob_threshold
         );
+
+        let mut blobs_to_delete = Vec::new();
         let mut wb = pending_db.batch();
-        for (k, _) in pending {
+
+        for (k, metadata) in pending {
+            blobs_to_delete.push(metadata.event_cursor.element_index);
             wb.delete_batch(pending_db, std::iter::once(k))?;
         }
-        if !attested.is_empty() {
+
+        for (_, metadata) in attested {
+            blobs_to_delete.push(metadata.event_cursor.element_index);
             wb.delete_batch(attested_db, std::iter::once(()))?;
         }
-        if !failed_to_attest.is_empty() {
+
+        for (_, metadata) in failed_to_attest {
+            blobs_to_delete.push(metadata.event_cursor.element_index);
             wb.delete_batch(failed_to_attest_db, std::iter::once(()))?;
         }
+
         wb.write()?;
+
+        for blob_index in blobs_to_delete {
+            let blob_path = blobs_path.join(blob_index.to_string());
+            if blob_path.exists() {
+                if let Err(e) = fs::remove_file(&blob_path) {
+                    tracing::warn!(?blob_path, ?e, "Failed to remove blob file");
+                } else {
+                    tracing::info!(?blob_path, "Removed uncertified blob file");
+                }
+            }
+        }
```

Comment:
question: what if the node stops during the removal? That may leave the blob file permanently on disk? Should we move the `wb.write()` after the removal, or do something like `ls` then `rm`?

Reply:
@halfprice I think it is dangerous to remove the blobs first before the db write. If node crashes after blob removal from filesystem but before db write, upon next restart the latest certified event blob could have progressed further to potentially include some of the blobs we are removing from file system (but they exist in db). I think one thing we can do to address your concern is, upon startup just check all blobs in db and filesystem, and remove ones from fs which are not in db.

Reply:
> upon next restart the latest certified event blob could have progressed further to potentially include some of the blobs we are removing from file system

Are you talking about the case where after restart, the node no longer needs to reset the event blob? And the removed event blobs can be potentially useful again?

> upon startup just check all blobs in db and filesystem, and remove ones from fs which are not in db.
Yeah, this sounds good to me too.

Reply:
> Are you talking about the case where after restart, the node no longer needs to reset the event blob? And the removed event blobs can be potentially useful again?

Yes